src/wcl/doc.h

```diff
@@ -62,10 +73,22 @@ struct doc_state {
   static doc_state from_string(const std::string& str) {
     doc_state state;
 
-    state.character_count = str.size();
+    state.byte_count = str.size();
+
+    utf8proc_int32_t codepoint;
+    utf8proc_ssize_t pos = 0;
+    while (true) {
+      pos += utf8proc_iterate(reinterpret_cast<const unsigned char*>(str.c_str()) + pos, -1,
```

Comment:
reinterpret_cast to `const utf8_proc_uint8_t*` once above and call it iter, then call

```
iter += utf8proc_iterate(iter, -1, &codepoint);
```

Also nit: declare `codepoint` inside the loop and the loop should be `while (iter < iter_end)` where iter_end is `iter + str.size()`. Both pos and codepoint are made negative so you can rely on codepoint to check for the error case.